src/nni_manager/training_service/dlts/dltsTrainingService.ts

```diff
@@ -0,0 +1,470 @@
+        const trialJob = this.trialJobsMap.get(trialJobId);
+        if (trialJob === undefined) {
+            throw Error(`Trial job ${trialJobId} not found.`)
+        }
+        return trialJob
+    }
+
+    public addTrialJobMetricListener(listener: (metric: TrialJobMetric) => void): void {
+        this.metricsEmitter.on('metric', listener);
+    }
+
+    public removeTrialJobMetricListener(listener: (metric: TrialJobMetric) => void): void {
+        this.metricsEmitter.off('metric', listener);
+    }
+
+    public get MetricsEmitter(): EventEmitter {
+        return this.metricsEmitter;
+    }
+
+    public async submitTrialJob(form: TrialJobApplicationForm): Promise<TrialJobDetail> {
+        const trialJobId: string = uniqueString(5);
+        const trialWorkingFolder: string = path.join(
+            '/nni/experiments', getExperimentId(),
+            '/trials/', trialJobId);
+        const trialJobDetail = new DLTSTrialJobDetail(
+            trialJobId, // id
+            'WAITING', // status
+            Date.now(), // submitTime
+            trialWorkingFolder, // workingDirectory
+            form,
+            `nni_exp_${this.experimentId}_trial_${trialJobId}`
+        );
+
+        this.trialJobsMap.set(trialJobId, trialJobDetail);
+        this.jobQueue.push(trialJobId);
+
+        return trialJobDetail;
+    }
+
+    public async cancelTrialJob(trialJobId: string, isEarlyStopped: boolean = false): Promise<void> {
+        const trialJobDetail: DLTSTrialJobDetail | undefined =  this.trialJobsMap.get(trialJobId);
+        if (trialJobDetail === undefined) {
+            throw Error(`cancelTrialJob: trial job id ${trialJobId} not found`);
+        }
+
+        if (this.dltsClusterConfig === undefined) {
+            throw Error('DLTS Cluster config is not initialized');
+        }
+
+        const options: request.Options = {
+            method: 'PUT',
+            uri: `${this.dltsClusterConfig.dashboard}api/clusters/${this.dltsClusterConfig.cluster}/jobs/${trialJobDetail.dltsJobId}/status`,
+            qs: {
+                email: this.dltsClusterConfig.email,
+                password: this.dltsClusterConfig.password
+            },
+            body: {
+                status: 'killing'
+            },
+            json: true
+        };
+
+        // Set trialjobDetail's early stopped field, to mark the job's cancellation source
+        trialJobDetail.isEarlyStopped = isEarlyStopped;
+
+        await new Promise((resolve, reject) => {
+            request(options, (error: Error, response: request.Response, body: any) => {
+                if (error) {
+                    reject(error);
+                } else {
+                    resolve(body);
+                }
+            });
+        });
+    }
+
+    public async setClusterMetadata(key: string, value: string): Promise<void> {
+        switch (key) {
+            case TrialConfigMetadataKey.NNI_MANAGER_IP:
```

Comment:
since nnimanager runs in a container in dlts, this ip is usually automatically detected by nni, or users have to set this IP manually. I means the practice which should be mentioned in doc.

Reply:
automatically detected.